src/GraphQL/Utilities/SchemaPrinter.cs

```diff
@@ -13,27 +13,28 @@ public class SchemaPrinter : IDisposable
     {
         protected SchemaPrinterOptions Options { get; }
 
-        private readonly List<string> _scalars = new List<string>(
-            new[]
-            {
-                "String",
-                "Boolean",
-                "Int",
-                "Float",
-                "ID"
-            });
+        private readonly List<string> _builtInScalars = new List<string>
+        {
+            "String",
+            "Boolean",
+            "Int",
+            "Float",
+            "ID"
+        };
+
+        private readonly List<string> _buildInDirectives = new List<string>
+        {
+            "skip",
+            "include",
+            "deprecated"
+        };
 
         public SchemaPrinter(
             ISchema schema,
             SchemaPrinterOptions options = null)
         {
             Schema = schema;
             Options = options ?? new SchemaPrinterOptions();
-
-            if (Options.CustomScalars?.Count > 0)
```

Comment:
The name of this property did not match its use.

Reply:
This option allows the user to add a custom scalar that will be treated as if it was a built-in scalar and never be printed.  Removing this removes that ability.

Reply:
I guessed about it. The question is **why** is such a _feature_ needed at all? Custom scalar is the same part of the schema along with all its other parts. What is the point of not printing this scalar in SDL? This scalar is used somewhere in the schema - in field arguments, in field types, etc. The resulting schema will simply be an incorrect document.

Reply:
The same reason why `String` is not printed.  If the type is considered part of your domain, there is no reason to print it.  That does not make it incorrect.

I could make the same argument you are about excluding scalar types from printing when they are not used.  They are defined in the Schema, you are just excluding them from printing because you don't want to see them.  Just because they are not used does not mean they don't exist.

Both of things are just features of the printer.

Reply:
> The same reason why String is not printed.

This is a different situation. The `String` is a built-in scalar. The specification clearly defines the behavior when displaying this scalar in SDL:
> When representing a GraphQL schema using the type system definition language, the built‐in scalar types **should be** omitted for brevity.

Regarding unused scalars. Do I understand correctly that you have already read my questions in the spec repo - graphql/graphql-spec#648 ?

>  Just because they are not used does not mean they don't exist.

Nope. This just means that they do not exist. The specification says this, but not as explicitly as we would like. In fact, for a long time, just no one paid attention to these scalars in SDL. If we add support for new scalar types, it doesn’t mean that these scalars should be printed in SDL.

In the end, the application developer does not even know about these scalars. This is just some payload from us as library developers to help the application developer. We provide an **initial set** of supported (native for .NET) scalars, but these should by no means **be equated** with the use of these scalars. These are two completely different things. In the future, we can add 100 new scalars, but this should not change the application schema.

Reply:
> This is a different situation.

No, "should be omitted for brevity" - is the exact reason why that feature is there.  So that is the exact same situation.

> Nope. This just means that they do not exist.

They do exist in the code, and the Schema configuration, they would just not being printed with this class.  They would still show up in introspection.

> Do I understand correctly that you have already read my questions in the spec repo

Yes.  Though adding this feature does not mean you have to remove a completely unrelated feature.

There are a ton of features that this codebase has that are not related to the spec at all.  This is one of them.  There is no reason to remove it as it is an optional feature and has its uses.

Reply:
> They would still show up in introspection.

So in this regard, if that is really the end goal here, I would say this filtering should actually be in the Schema itself and not just this printer.

Reply:
Indeed, introspection needs to be fixed.

Reply:
By the way, I remembered that just in case I made the `SchemaPrinterOptions.IncludeAllTypes` property. But to be honest, I still don’t understand when it can be useful to form an SDL in which there will be custom scalars that are not actually used in the SDL.

Reply:
> I still don’t understand when it can be useful to form an SDL in which there will be custom scalars that are not actually used in the SDL.

When me, as a human, wants to view the printed schema without all the extra scalars.

Reply:
Looking at this class as a whole again, it does appear there are enough hooks in place to accomplish not printing the scalars without this feature in the Options.  I'm okay with removing it.

I _do_ think that the filtering should actually be done on the Schema itself though and not in the printer.  That would give us some hooks to also support filtering the Schema based on Authentication/Authorization rules.

Reply:
Done for introspection.

> I do think that the filtering should actually be done on the Schema itself though and not in the printer.

Something like `ISchema.AllUsedTypes` ? I think it will look strange.

Reply:
> When me, as a human, wants to view the printed schema without all the extra scalars.

So... It seemed to me that I was doing just that.

Reply:
This feature will exclude _some_ scalars, but not _all_ scalars.

Reply:
I would probably do the filtering in this property.

https://github.com/graphql-dotnet/graphql-dotnet/blob/424f59696fa82396ea0b74013804b4793aff8d56/src/GraphQL/Types/Schema.cs#L90

... and provide a way to control the filtering.  Right now there would be no way to turn off the reference check, and it is hard-coded in both the printer and introspection.  You can instead move the filtering to the Schema and not have to make changes to the printer nor the introspection types.

I added this `Filter` feature, but it is only used in Introspection today.

https://github.com/graphql-dotnet/graphql-dotnet/blob/424f59696fa82396ea0b74013804b4793aff8d56/src/GraphQL/Types/Schema.cs#L74

This idea could be applied more broadly to all of the types in the Schema.

Reply:
If I do filtering in `AllTypes` property, it will most likely lead to errors, because the property is used everywhere to get the **full set** of types. This property is for internal GraphQL server consumption.

Reply:
One more thought about filtering. I understand that the user can filter parts of the schema. I do not consider this an unnecessary featurey. But I believe that it should be limited as much as possible so that as a result of this filtering a valid schema is obtained. If the specification says that the introspection request MUST return only used scalars, then only they should fall into the user filter, and not all scalars.